### PR TITLE
0003 character control

### PR DESCRIPTION
**Pull request by SallahAdDin**

### CODE REVIEW

File: Assets/Scripts/CameraControl.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+using System.Collections;
+
+public class CameraControl : MonoBehaviour {
+    [SerializeField]
+    Transform target;
+    [SerializeField]
+    float speed = 10;
+    [SerializeField]
+    float stopDistance;
+	// Update is called once per frame
```
Comment (semensan) [inline Assets/Scripts/CameraControl.cs]: Not relevant comment.
Delete it.

File: Assets/Scripts/CameraControl.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	void FixedUpdate () {
+        Vector3 projectedTarget = Vector3.ProjectOnPlane(target.transform.position, Vector3.up);
+        Vector3 projectedPosition = Vector3.ProjectOnPlane(transform.position, Vector3.up);
```
Comment (semensan) [inline Assets/Scripts/CameraControl.cs]: No need to project. Just y = 0.

File: Assets/Scripts/CameraControl.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        Vector3 direction = projectedTarget - projectedPosition;
+        Vector3 move = speed * (direction).sqrMagnitude * Time.fixedDeltaTime * direction.normalized;
+        if (direction.sqrMagnitude > stopDistance * stopDistance)
```
Comment (semensan) [inline Assets/Scripts/CameraControl.cs]: Possible optimization - store sqrStopDistance

File: Assets/Scripts/Gate.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using UnityEngine;
+using System.Collections;
+using UnityEngine.SceneManagement;
```
Comment (semensan) [inline Assets/Scripts/Gate.cs]: Sort libraries

File: Assets/Scripts/Gate.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+public class Gate : InteractableEnvironment {
+    [SerializeField]
+    string targetSceneName;
+
+    public override void  OnInteraction()
+    {
+        Debug.Log("Open");
```
Comment (semensan) [inline Assets/Scripts/Gate.cs]: And what will you do when log file will be 100500 lines long?
Write understandable description

File: Assets/Scripts/InteractableEnvironment.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public abstract class InteractableEnvironment : MonoBehaviour {
+    
+    [SerializeField]
+    Transform interactPoint;
+    public Vector3 interactPosition
```
Comment (semensan) [inline Assets/Scripts/InteractableEnvironment.cs]: New line here

File: Assets/Scripts/InteractableEnvironment.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public abstract class InteractableEnvironment : MonoBehaviour {
+
```
Comment (semensan) [inline Assets/Scripts/InteractableEnvironment.cs]: Delete new line

File: Assets/Scripts/InteractableEnvironment.cs (hunk 6 above)
Comment (semensan) [inline Assets/Scripts/InteractableEnvironment.cs]: Complicated logic = function, not field

File: Assets/Scripts/PlayerMouseControl.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
+﻿using UnityEngine;
+using System.Collections;
+[RequireComponent(typeof(NavMeshAgent))]
```
Comment (semensan) [inline Assets/Scripts/PlayerMouseControl.cs]: New line here

File: Assets/Scripts/PlayerMouseControl.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
+﻿using UnityEngine;
+using System.Collections;
```
Comment (semensan) [inline Assets/Scripts/PlayerMouseControl.cs]: New line below this

File: Assets/Scripts/PlayerMouseControl.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+[RequireComponent(typeof(NavMeshAgent))]
+public class PlayerMouseControl : MonoBehaviour {
+    public Vector3 target;
+    NavMeshAgent agent;
+    public float speed = 10;
+    InteractableEnvironment toInteract;
+	// Use this for initialization
```
Comment (semensan) [inline Assets/Scripts/PlayerMouseControl.cs]: Delete comment and make one indentation

File: Assets/Scripts/PlayerMouseControl.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	void Start () {
+        agent = GetComponent<NavMeshAgent>();
+        target = transform.position;
+        agent.SetDestination(target);
+	}
+	
+	// Update is called once per frame
```
Comment (semensan) [inline Assets/Scripts/PlayerMouseControl.cs]: Delete comment

**Comment by semensan** [comment]
Rename commit "Add interact mechanic" to "Add interaction mechanic"
"Add city enter as interact Example" to "Add city entrance interaction"
Удатить Terrain из несвязанных коммитов

File: Assets/Scripts/CameraControl.cs (hunk 1 above)
Comment (SallahAdDin) [inline Assets/Scripts/CameraControl.cs]: +

File: Assets/Scripts/CameraControl.cs [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+public class CameraControl : MonoBehaviour {
+    [SerializeField]
+    Transform target;
+    [SerializeField]
+    float speed = 10;
+    [SerializeField]
+    float stopDistance;
+	// Update is called once per frame
+	void FixedUpdate () {
```
Comment (SallahAdDin) [inline Assets/Scripts/CameraControl.cs]: +

File: Assets/Scripts/CameraControl.cs (hunk 2 above)
Comment (SallahAdDin) [inline Assets/Scripts/CameraControl.cs]: +

File: Assets/Scripts/CameraControl.cs (hunk 3 above)
Comment (SallahAdDin) [inline Assets/Scripts/CameraControl.cs]: Точно оптимизация?

File: Assets/Scripts/Gate.cs (hunk 4 above)
Comment (SallahAdDin) [inline Assets/Scripts/Gate.cs]: +

File: Assets/Scripts/Gate.cs (hunk 5 above)
Comment (SallahAdDin) [inline Assets/Scripts/Gate.cs]: +

File: Assets/Scripts/InteractableEnvironment.cs (hunk 7 above)
Comment (SallahAdDin) [inline Assets/Scripts/InteractableEnvironment.cs]: +

File: Assets/Scripts/InteractableEnvironment.cs (hunk 6 above)
Comment (SallahAdDin) [inline Assets/Scripts/InteractableEnvironment.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs (hunk 9 above)
Comment (SallahAdDin) [inline Assets/Scripts/PlayerMouseControl.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs (hunk 10 above)
Comment (SallahAdDin) [inline Assets/Scripts/PlayerMouseControl.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs (hunk 11 above)
Comment (SallahAdDin) [inline Assets/Scripts/PlayerMouseControl.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+[RequireComponent(typeof(NavMeshAgent))]
+public class PlayerMouseControl : MonoBehaviour 
+{
+    public Vector3 target;
+    NavMeshAgent agent;
+    public float speed = 10;
+    InteractableEnvironment toInteract;
+
+	void Start () 
+    {
```
Comment (semensan) [inline Assets/Scripts/PlayerMouseControl.cs]: Respect coding style

File: Assets/Scripts/PlayerMouseControl.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        agent = GetComponent<NavMeshAgent>();
+        target = transform.position;
+        agent.SetDestination(target);
+	}
+	
+	void FixedUpdate () 
+    {
```
Comment (semensan) [inline Assets/Scripts/PlayerMouseControl.cs]: Coding style + Input events in Update

File: Assets/Scripts/Walkable.cs [hunk 15]
```diff
@@ -0,0 +1,7 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public class Walkable : MonoBehaviour 
+{
+
```
Comment (semensan) [inline Assets/Scripts/Walkable.cs]: No new line?

File: Assets/Scripts/CameraControl.cs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+public class CameraControl : MonoBehaviour 
+{
+    [SerializeField]
+    Transform target;
+    [SerializeField]
+    float speed = 10;
+    [SerializeField]
+    float stopDistance;
+    float sqrStopDistance;
```
Comment (SallahAdDin) [inline Assets/Scripts/CameraControl.cs]: +

File: Assets/Scripts/CameraControl.cs [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+	void Update () 
+    {
+        Vector3 projectedTarget = target.transform.position;
+        projectedTarget.Set(projectedTarget.x, 0, projectedTarget.z);
+        Vector3 projectedPosition = transform.position;
+        projectedPosition.Set(projectedPosition.x, 0, projectedPosition.z);
+        Vector3 direction = projectedTarget - projectedPosition;
+        Vector3 move = speed * (direction).sqrMagnitude * Time.fixedDeltaTime * direction.normalized;
```
Comment (SallahAdDin) [inline Assets/Scripts/CameraControl.cs]: +

File: Assets/Scripts/Gate.cs [hunk 18]
```diff
@@ -0,0 +1,14 @@
+﻿using UnityEngine;
+using UnityEngine.SceneManagement;
+using System.Collections;
+
+public class Gate : InteractableEnvironment 
+{
+    [SerializeField]
+    string targetSceneName;
+
+    public override void  OnInteraction()
```
Comment (SallahAdDin) [inline Assets/Scripts/Gate.cs]: Let it be one line
reject

File: Assets/Scripts/InteractableEnvironment.cs [hunk 19]
```diff
@@ -0,0 +1,21 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public abstract class InteractableEnvironment : MonoBehaviour 
+{
+    
+    [SerializeField]
```
Comment (SallahAdDin) [inline Assets/Scripts/InteractableEnvironment.cs]: +

File: Assets/Scripts/InteractableEnvironment.cs [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+    Transform interactPoint;
+    public Vector3 GetInteractPosition()
```
Comment (SallahAdDin) [inline Assets/Scripts/InteractableEnvironment.cs]: +

File: Assets/Scripts/InteractableEnvironment.cs [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+    {
+            NavMeshHit nmHit;
+            if (NavMesh.SamplePosition(interactPoint.transform.position, out nmHit, Mathf.Infinity, NavMesh.AllAreas))
+            {
+                return nmHit.position;
+            }
+            return Vector3.zero;
+    }
+    public float interactDistance;
```
Comment (SallahAdDin) [inline Assets/Scripts/InteractableEnvironment.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs [hunk 22]
```diff
@@ -0,0 +1,71 @@
+﻿using UnityEngine;
+using System.Collections;
+
+[RequireComponent(typeof(NavMeshAgent))]
+public class PlayerMouseControl : MonoBehaviour 
+{
+    public Vector3 target;
+    NavMeshAgent agent;
+    public float speed = 10;
+    InteractableEnvironment toInteract;
```
Comment (SallahAdDin) [inline Assets/Scripts/PlayerMouseControl.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs (hunk 13 above)
Comment (SallahAdDin) [inline Assets/Scripts/PlayerMouseControl.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs (hunk 14 above)
Comment (SallahAdDin) [inline Assets/Scripts/PlayerMouseControl.cs]: +

File: Assets/Scripts/Walkable.cs (hunk 15 above)
Comment (SallahAdDin) [inline Assets/Scripts/Walkable.cs]: +

File: Assets/Scripts/InteractableEnvironment.cs [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+
+    public Vector3 GetInteractPosition()
+    {
+            NavMeshHit nmHit;
+            if (NavMesh.SamplePosition(interactPoint.transform.position, out nmHit, Mathf.Infinity, NavMesh.AllAreas))
+            {
+                return nmHit.position;
+            }
+            return Vector3.zero;
```
Comment (semensan) [inline Assets/Scripts/InteractableEnvironment.cs]: Two tabs?

File: Assets/Scripts/InteractableEnvironment.cs [hunk 24]
```diff
@@ -0,0 +1,21 @@
+﻿using UnityEngine;
+using System.Collections;
+
+public abstract class InteractableEnvironment : MonoBehaviour 
+{
+    [SerializeField]
+    Transform interactPoint;
+    public float interactDistance;
```
Comment (semensan) [inline Assets/Scripts/InteractableEnvironment.cs]: public first

File: Assets/Scripts/PlayerMouseControl.cs [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+[RequireComponent(typeof(NavMeshAgent))]
+public class PlayerMouseControl : MonoBehaviour 
+{
+    public Vector3 target;
+    public float speed = 10;
+    NavMeshAgent agent;
+    InteractableEnvironment toInteract;
+
+	void Start () 
+    {
```
Comment (semensan) [inline Assets/Scripts/PlayerMouseControl.cs]: fix indentation

File: Assets/Scripts/InteractableEnvironment.cs (hunk 24 above)
Comment (SallahAdDin) [inline Assets/Scripts/InteractableEnvironment.cs]: +

File: Assets/Scripts/InteractableEnvironment.cs (hunk 23 above)
Comment (SallahAdDin) [inline Assets/Scripts/InteractableEnvironment.cs]: +

File: Assets/Scripts/PlayerMouseControl.cs (hunk 25 above)
Comment (SallahAdDin) [inline Assets/Scripts/PlayerMouseControl.cs]: +

**Pull request merged**